lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -139,95 +150,275 @@
           }
         }
 '''
+import json
+
+from abc import ABCMeta, abstractmethod
+
+from ansible.module_utils._text import to_native, to_text
+from ansible.module_utils.basic import AnsibleModule, get_all_subclasses
+from ansible.module_utils.six import with_metaclass
+
+
+class TestFailed(Exception):
+    pass
+
+
+class PkgMgr(with_metaclass(ABCMeta, object)):
+
+    def __init__(self):
+
+        if not self.test():
+            raise TestFailed("Test failed to pass")
+
+    @abstractmethod
+    def test(self):
+        ''' This method is supposed to return True/False if the package manager is installed/usable by this module '''
+        pass
+
+    @abstractmethod
+    def list_installed(self):
+        ''' This method should return a list of installed packages, each list item will be passed to get_package_details '''
+        pass
+
+    @abstractmethod
+    def get_package_details(self, package):
+        ''' This takes a 'package' item and returns a dictionary with the package information, name and version are minimal requirements '''
+        pass
+
+    def get_packages(self):
+        ''' Take all of the above and return a dictionary of lists of dictionaries (package = list of installed versions) '''
+
+        installed_packages = {}
+        for package in self.list_installed():
+            package_details = self.get_package_details(package)
+            package_details['source'] = self.__class__.__name__.lower()
+            name = package_details['name']
+            if name not in installed_packages:
+                installed_packages[name] = [package_details]
+            else:
+                installed_packages[name].append(package_details)
+        return installed_packages
+
+
+class LibMgr(PkgMgr):
+
+    LIB = None
+
+    def __init__(self):
+
+        self.lib = None
+        super(LibMgr, self).__init__()
+
+    def test(self):
+        found = False
+        try:
+            self.lib = __import__(self.LIB)
+            found = True
+        except ImportError:
+            pass
+        return found
```

Comment:
It probably even makes more sense to put exception raising into this function.
Otherwise it's a three-step process on the high level: first, call the method, then, check the result, finally raise an exception.

It would be much cleaner to have smth like `self.verify_cli_backend_available()` in the initializer and that's it.

Reply:
+1